### PR TITLE
Add Nuxt and Vue Router properties in `order-in-components`

### LINKED ISSUE #980: Add all Nuxt properties in order-in-components
Nuxt.js's [asyncData](https://nuxtjs.org/api/), [fetch](https://nuxtjs.org/api/pages-fetch) and [head](https://nuxtjs.org/api/pages-head) component properties were already added to the `order-in-components` rule.

There are others, though:

* [key](https://nuxtjs.org/api/pages-key)
* [layout](https://nuxtjs.org/api/pages-layout)
* [loading](https://nuxtjs.org/api/pages-loading)
* [middleware](https://nuxtjs.org/api/pages-middleware)
* [scrollToTop](https://nuxtjs.org/api/pages-scrolltotop)
* [transition](https://nuxtjs.org/api/pages-transition)
* [validate](https://nuxtjs.org/api/pages-validate)
* [watchQuery](https://nuxtjs.org/api/pages-watchquery)

Additionally, Vue Router defines three [in-component navigation guards](https://router.vuejs.org/guide/advanced/navigation-guards.html#in-component-guards):

* beforeRouteEnter
* beforeRouteUpdate
* beforeRouteLeave

It would be useful to have them included in the default rule config to enforce a consistent style.

I'd like to suggest the following order:

```diff
   // side effects
   "el",

   // global awareness
   "name",
+  "key",
   "parent",

   // component type
   "functional",

   // template modifiers
   ["delimiters", "comments"],

   // template dependencies
   ["components", "directives", "filters"],

   // composition
   "extends",
   "mixins",

+  // new category: page validation
+  "beforeRouteEnter",
+  "beforeRouteUpdate",
+  "beforeRouteLeave",
+  "middleware",
+  "validate",

   // interface
   "inheritAttrs",
   "model",
   ["props", "propsData"],

   // local state
   "fetch",
   "asyncData",
   "data",
   "computed",

   // events
   "watch",
+  "watchQuery",
   "LIFECYCLE_HOOKS",

   // non-reactive properties
   "methods",
   "head",

   // rendering
+  "layout",
   ["template", "render"],
   "renderError",
+  "transition",
+  "loading",
+  "scrollToTop",

```

### PR DESCRIPTION
**Pull request by FloEdelmann**
Closes #980.

### CODE REVIEW

**Comment by ota-meshi** [comment]
Thank you for this PR!

I think it's good to add Nuxt properties as standard.
But I'm not sure if this order is acceptable to Nuxt developers. 
Is there a documented recommended Nuxt property order?

File: lib/rules/order-in-components.js [hunk 1]
```diff
@@ -8,27 +8,59 @@ const utils = require('../utils')
...
+  'head', // Nuxt
+
+  // rendering
+  'layout', // Nuxt
   ['template', 'render'],
-  'renderError'
+  'renderError',
+  'transition', // Nuxt
+  'loading', // Nuxt
+  'scrollTop' // Nuxt
```
Comment (ota-meshi) [inline lib/rules/order-in-components.js:63]: `scrollToTop`?

**Comment by FloEdelmann** [comment]
@ota-meshi Unfortunately, as far as I know, there isn't a recommended property order.

@Atinux from the Nuxt.js core team: Are there plans to add a style guide similar to Vue's to the Nuxt documentation? Is the order suggested in this PR suitable?

**Comment by FloEdelmann** [comment]
@ota-meshi I updated the pull request to reflect your changes for Vue 3. Unfortunately, none of the Nuxt maintainers has responded to any of the issues yet :slightly_frowning_face:

**Comment by ota-meshi** [comment]
@FloEdelmann Thank you for update.

Hi @manniL.
Can you review if this order is common to Next developers?

File: docs/rules/order-in-components.md [hunk 2]
```diff
@@ -65,31 +65,48 @@ export default {
...
       "fetch",
       "asyncData",
       "data",
       "computed",
       "watch",
+      "watchQuery",
       "LIFECYCLE_HOOKS",
       "methods",
       "head",
+      "layout",
```
Comment (Atinux) [inline docs/rules/order-in-components.md:93]: From my POV, I believe that Nuxt properties should be at least before `methods`,  I don't want to scroll til the end to see what special properties my page has

Comment (FloEdelmann) [reply docs/rules/order-in-components.md:93]: @Atinux Thanks for your answer. What do you think about the following?

```diff
  // Global Awareness (requires knowledge beyond the component)
  'name',
+ 'key', // for Nuxt
  'parent',

  // Component Type (changes the type of the component)
  'functional',

  // Template Modifiers (changes the way templates are compiled)
  ['delimiters', 'comments'],

  // Template Dependencies (assets used in the template)
  ['components', 'directives', 'filters'],

  // Composition (merges properties into the options)
  'extends',
  'mixins',
  ['provide', 'inject'], // for Vue.js 2.2.0+

- // Page Validation (determines if the router request is suitable)
+ // Page Options (component rendered as a router page)
+ 'ROUTER_GUARDS', // for Vue Router
+ 'middleware', // for Nuxt
+ 'validate', // for Nuxt
+ 'layout', // for Nuxt
+ 'transition', // for Nuxt
+ 'loading', // for Nuxt
+ 'scrollToTop', // for Nuxt

  // Interface (the interface to the component)
  …

  // Local State (local reactive properties)
  …

  // Events (callbacks triggered by reactive events)
  'watch',
+ 'watchQuery', // for Nuxt
  'LIFECYCLE_HOOKS',

  // Non-Reactive Properties (instance properties independent of the reactivity system)
  'methods',

  // Rendering (the declarative description of the component output)
  'head', // for Nuxt
  ['template', 'render'],
  'renderError',
```

Comment (Atinux) [reply docs/rules/order-in-components.md:93]: After talking with @alexchopin and @pi0, we suggest the following:

```js
{
  "vue/order-in-components": ["error", {
    "order": [
      "el",
      "name",
      "key",
      "parent",
      "functional",
      ["delimiters", "comments"],
      ["components", "directives", "filters"],
      "extends",
      "mixins",
      ["provide", "inject"],
      "ROUTER_GUARDS",
      "layout",
      "middleware",
      "validate",
      "scrollToTop",
      "transition",
      "loading",
      "inheritAttrs",
      "model",
      ["props", "propsData"],
      "emits",
      "setup",
      "fetch",
      "asyncData",
      "data",
      "head",
      "computed",
      "watch",
      "watchQuery",
      "LIFECYCLE_HOOKS",
      "methods",
      ["template", "render"],
      "renderError",
    ]
  }]
```

Comment (FloEdelmann) [reply docs/rules/order-in-components.md]: Thanks, I've updated the PR to reflect the suggested order (0aa0765).

@ota-meshi Is this order acceptable for you?

**Pull request merged**